outlook.com: iPadOS: Have to tap twice to change text highlight / color when composing an email (overeager content change observer)

megangardner opened the pull request:
#### 0997d278de154aa54b5772f00c4d3b75b1c8363d
<pre>
outlook.com: iPadOS: Have to tap twice to change text highlight / color when composing an email (overeager content change observer)
<a href="https://bugs.webkit.org/show_bug.cgi?id=294059">https://bugs.webkit.org/show_bug.cgi?id=294059</a>
<a href="https://rdar.apple.com/152277211">rdar://152277211</a>

Reviewed by Brent Fulgham.

The issue here is that the first tap on the color triggers a Content Change Observation-based *hover*.
You have to tap on the color again to actually commit to the color change, and then all is well.
Therefore we should ignore content changed observer for these specific elements so that their changes
will propagate to the page permanently.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldDispatchSimulatedMouseEvents const):
(WebCore::Quirks::shouldIgnoreContentObservationForClick const):
(WebCore::handleLiveQuirks):

Canonical link: <a href="https://commits.webkit.org/295923@main">https://commits.webkit.org/295923@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47c23c585252e4268c8ce46f1b35511d32edd1f3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106521 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26272 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16669 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111743 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57117 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26942 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34773 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80901 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109525 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21354 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96115 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61236 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20815 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14217 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56567 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90692 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14251 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114583 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33659 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24807 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89959 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34023 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92346 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89669 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/22884 "The change is no longer eligible for processing.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34567 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12384 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29268 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17266 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33584 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38997 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33330 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36683 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34929 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1687,6 +1687,9 @@ bool Quirks::shouldIgnoreContentObservationForClick(const Node& targetNode) cons
         return false;
 
     RefPtr target = dynamicDowncast<Element>(targetNode);
+    if (target->getIdAttribute().startsWith("swatchColorPicker"_s))
```
this will be a crash if target is null

pxlcoder replied:
Additionally, I think the way this is written we'd also check for `swatchColorPicker` on walmart.com?

karlcow replied:
probably `if (target && target->getIdAttribute().startsWith("swatchColorPicker"_s))`

karlcow replied:
> Additionally, I think the way this is written we'd also check for `swatchColorPicker` on walmart.com?

Yes see my comment on https://github.com/WebKit/WebKit/pull/46357#discussion_r2127813579

pxlcoder commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -433,7 +433,6 @@ bool Quirks::shouldDispatchSimulatedMouseEvents(const EventTarget* target) const
                 return true;
         }
         return false;
-
```
Please undo.

karlcow commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1678,6 +1677,7 @@ static AccessibilityRole accessibilityRole(const Element& element)
 
 // walmart.com: rdar://123734840
+// live.outlook.com: rdar://152277211
 bool Quirks::shouldIgnoreContentObservationForClick(const Node& targetNode) const
```
This now applies to two different domains with different targets and attributes. The code needs to isolate the scope per domain to avoid name clashes. 

(Non-Reviewer status review) 
I let @brentfulgham chime in for a full review or @whsieh

karlcow replied:
To avoid this you can do something like:
https://github.com/WebKit/WebKit/blob/6c66a43574f55efadbca8e69b88161d541b36d10/Source/WebCore/page/Quirks.cpp#L983-L1000

karlcow replied:
Maybe better if we have two OutLook Quirks. I can refactor later if you want. 
https://github.com/WebKit/WebKit/blob/6c66a43574f55efadbca8e69b88161d541b36d10/Source/WebCore/page/Quirks.cpp#L832C1-L842

karlcow reviewed (commented):
That looks good to me. Thanks a lot for this nice refactor @megangardner 

Let's see if @whsieh or @brentfulgham have additional comments or can approve.

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -2469,6 +2478,11 @@ static void handleLiveQuirks(QuirksData& quirksData, const URL& quirksURL, const
     auto topDocumentHost = quirksURL.host();
     // outlook.live.com: rdar://136624720
     quirksData.needsMozillaFileTypeForDataTransferQuirk = topDocumentHost == "outlook.live.com"_s;
```
Rather than do the string comparison three times (on iOS), maybe we could set the `quirksData.isOutlook` first (on all platforms), and then use that value to set `quirksData.needsMozillaFileTypeForDataTransferQuirk` (on all platforms), and `mayNeedToIgnoreContentObservation` on iOS?

brentfulgham reviewed (approved):
Looks good to me, if you avoid the three identical string comparisons in `handleLiveQuirks` as I suggest.

r=mews otherwise, though!

megangardner committed:
outlook.com: iPadOS: Have to tap twice to change text highlight / color when composing an email (overeager content change observer)
https://bugs.webkit.org/show_bug.cgi?id=294059
rdar://152277211

Reviewed by Brent Fulgham.

The issue here is that the first tap on the color triggers a Content Change Observation-based *hover*.
You have to tap on the color again to actually commit to the color change, and then all is well.
Therefore we should ignore content changed observer for these specific elements so that their changes
will propagate to the page permanently.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldDispatchSimulatedMouseEvents const):
(WebCore::Quirks::shouldIgnoreContentObservationForClick const):
(WebCore::handleLiveQuirks):

Canonical link: https://commits.webkit.org/295923@main
